File: src/main/java/com/topic/enumerate/NumberOfPairs.java
package com.topic.enumerate;

import org.junit.jupiter.api.Test;

import java.util.HashMap;

/**
 * @author zc.zhou
 * @Description 3162. 优质数对的总数 I 枚举
 * @create 2024-10-10 8:49
 */
public class NumberOfPairs {

  //o(n²)
  public int numberOfPairsV1(int[] nums1, int[] nums2, int k) {
    int res = 0;
    for (int n1 : nums1) {
      for (int n2 : nums2) {
        if (n1 % (n2 * k) == 0) {
          res++;
        }
      }
    }
    return res;
  }


  @Test
  void testFun() {
    int[] nums1 = {1,3,4};
    int[] nums2 = {1,3,4};
    System.out.println(numberOfPairs(nums1,nums2,1));
  }

  //o(n²)
  public long numberOfPairs(int[] nums1, int[] nums2, int k) {
    //分别统计nums1和nums2中每个数出现的次数
    HashMap<Integer, Integer> cntMap1 = new HashMap<>();
    //nums1最大值
    int max1 = Integer.MIN_VALUE;
    HashMap<Integer, Integer> cntMap2 = new HashMap<>();
    for(int n1 : nums1) {
      cntMap1.put(n1, cntMap1.getOrDefault(n1, 0) + 1);
      max1 = Math.max(max1, n1);
    }
    for(int n2 : nums2) {
      cntMap2.put(n2, cntMap2.getOrDefault(n2, 0) + 1);
    }
    long res = 0;
    //枚举nums2中每一个数
    for (int n2 : cntMap2.keySet()) {
      long tmp = (long) n2 * k;
      //i为倍数
      for (int i = 1; i * tmp <= max1 ; i++) {
        int tar = i * (int) tmp;
        res += (long) cntMap1.getOrDefault(tar, 0) * cntMap2.getOrDefault(n2, 0);
      }
    }
    return res;
  }
}

File: pyproject/enumerate/NumberOfPairs.py
from collections import Counter
from typing import List


class Solution:
    # 3164. 优质数对的总数 II 枚举
    def numberOfPairs(self, nums1: List[int], nums2: List[int], k: int) -> int:
        map1 = Counter(nums1)
        map2 = Counter(nums2)
        max1 = max(nums1)
        # 枚举倍数
        res = 0
        for n2, cnt2 in map2.items():
            tmp = k * n2
            for i in range(1, max1 // tmp + 1):  # i 从 1 开始，到 max1//tmp 为止
                res += map1.get(i * tmp, 0) * cnt2
        return res

    def numberOfPairsV1(self, nums1: List[int], nums2: List[int], k: int) -> int:
        map1 = {}
        map2 = {}
        max1 = nums1[0]
        for n1 in nums1:
            map1[n1] = map1.get(n1, 0) + 1
            max1 = max(max1, n1)
        for n2 in nums2:
            map2[n2] = map2.get(n2, 0) + 1
        # 枚举倍数
        res = 0
        for n2 in map2.keys():
            tmp = k * n2
            for i in range(1, max1 // tmp + 1):  # i 从 1 开始，到 max1//tmp 为止
                tar = i * tmp
                if tar <= max1:
                    res += map1.get(tar, 0) * map2.get(n2, 0)
        return res


solution = Solution()
num1 = [1, 3, 4]
num2 = [1, 3, 4]
print(solution.numberOfPairs(num1, num2, 1))
